evm/src/cross_table_lookup.rs

```diff
@@ -59,54 +74,90 @@ impl<F: Field> Column<F> {
         }
     }
 
-    pub fn linear_combination<I: IntoIterator<Item = (usize, F)>>(iter: I) -> Self {
+    pub(crate) fn linear_combination<I: IntoIterator<Item = WeightedColumn<F>>>(iter: I) -> Self {
         Self::linear_combination_with_constant(iter, F::ZERO)
     }
 
     pub fn le_bits<I: IntoIterator<Item = usize>>(cs: I) -> Self {
-        Self::linear_combination(cs.into_iter().zip(F::TWO.powers()))
+        Self::linear_combination(cs.into_iter().zip(F::TWO.powers()).map(|(column, weight)| {
+            WeightedColumn {
+                column,
+                next: true,
+                weight,
+            }
+        }))
     }
 
     pub fn sum<I: IntoIterator<Item = usize>>(cs: I) -> Self {
-        Self::linear_combination(cs.into_iter().zip(repeat(F::ONE)))
+        Self::linear_combination(cs.into_iter().map(|column| WeightedColumn {
+            column,
+            next: true,
+            weight: F::ONE,
+        }))
     }
 
-    pub fn eval<FE, P, const D: usize>(&self, v: &[P]) -> P
+    pub fn eval<FE, P, const D: usize>(&self, local_values: &[P], next_values: &[P]) -> P
     where
         FE: FieldExtension<D, BaseField = F>,
         P: PackedField<Scalar = FE>,
     {
         self.linear_combination
             .iter()
-            .map(|&(c, f)| v[c] * FE::from_basefield(f))
+            .map(|weighted_col| {
+                let values = if weighted_col.next {
+                    next_values
+                } else {
+                    local_values
+                };
+                values[weighted_col.column] * FE::from_basefield(weighted_col.weight)
+            })
             .sum::<P>()
             + FE::from_basefield(self.constant)
     }
 
     /// Evaluate on an row of a table given in column-major form.
-    pub fn eval_table(&self, table: &[PolynomialValues<F>], row: usize) -> F {
+    pub fn eval_table(&self, table: &[PolynomialValues<F>], local_row: usize) -> F {
+        let mut next_row = local_row + 1;
+        if next_row == table[0].len() {
+            next_row = 0;
+        }
```

Comment:
nit: Wouldn't `let next_row = (local_row + 1) % table[0].len()` be clearer?

Reply:
I agree it's clearer, but I think the compiler would resort to the CPU's division instructions, since it wouldn't know that `local_row` is smaller than the table or that it's a power of two. Do you think `& (table[0].len() - 1)` would be any clearer?

Reply:
Oh I see. Using `&` could be clearer along with a small comment reminding that `table[0].len()` is a power of 2, but otherwise I think it's totally fine to keep the conditional.